go/cmd/dolt/commands/checkout.go

```diff
@@ -84,28 +85,46 @@ func (cmd CheckoutCmd) Exec(ctx context.Context, commandStr string, args []strin
 	helpPrt, usagePrt := cli.HelpAndUsagePrinters(cli.CommandDocsForCommandString(commandStr, checkoutDocs, ap))
 	apr := cli.ParseArgsOrDie(ap, args, helpPrt)
 
-	queryEngine, sqlCtx, close, err := cliCtx.QueryEngine(ctx)
+	queryEngine, sqlCtx, closeFunc, err := cliCtx.QueryEngine(ctx)
 	if err != nil {
 		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usagePrt)
 	}
-	defer close()
+	defer closeFunc()
 
 	branchOrTrack := apr.Contains(cli.CheckoutCoBranch) || apr.Contains(cli.TrackFlag)
 	if (branchOrTrack && apr.NArg() > 1) || (!branchOrTrack && apr.NArg() == 0) {
 		usagePrt()
 		return 1
 	}
 
+	var branchName string
+	if apr.Contains(cli.CheckoutCoBranch) {
+		branchName, _ = apr.GetValue(cli.CheckoutCoBranch)
+	} else if apr.NArg() > 0 {
+		branchName = apr.Arg(0)
+	}
+
+	currentBranch, err := getActiveBranchName(sqlCtx, queryEngine)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usagePrt)
+	}
+
+	if branchName == currentBranch {
+		cli.Printf("Already on branch '%s'\n", branchName)
+		return 0
+	}
+
 	sqlQuery, err := generateCheckoutSql(args)
 	if err != nil {
 		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usagePrt)
 	}
 
 	_, err = getRowsForSql(queryEngine, sqlCtx, sqlQuery)
 
-	var branchName string
-	if apr.NArg() > 0 {
-		branchName = apr.Arg(0)
+	// This command doesn't modify `dEnv` which could break tests that call multiple commands in sequence.
+	// We must reload it so that it includes changes to the repo state.
+	if dEnv != nil {
```

Comment:
If we migrate every command, we'll be able to remove this.

This only exists because without it, when a go test runs a CliContext command followed by a DoltEnv command, the DoltEnv command has stale information about the repo state. Once there are no more DoltEnv commands, this problem goes away.